lib/testobject.js

```diff
@@ -0,0 +1,164 @@
+  }
+
+  // Determine the TestObject_device
+  let testobject_device;
+  if (process.env.TESTOBJECT_DEVICE) {
+    testobject_device = process.env.TESTOBJECT_DEVICE;
+  } else if (caps.platformName && caps.platformName.toLowerCase() === 'ios') {
+    testobject_device = TestObject.DEFAULT_IOS_DEVICE;
+  } else {
+    testobject_device = TestObject.DEFAULT_DEVICE;
+  }
+
+  // Extend the provided caps with special testobject caps 
+  // (see https://app.staging.testobject.org/#/danielgraham/contact-manager/appium/basic/instructions)
+  return  Object.assign(caps, {
+    testobject_app_id,
+    testobject_device,
+    testobject_appium_version: '1.6.0',
+    testobject_api_key: process.env.TESTOBJECT_API_KEY,
+    testobject_cache_device: true,
+  });
+};
+
+TestObject._appIdCache = {};
+
+// Make request-promise mockable
+TestObject.request = request;
+
+/**
+ * Uploads your app (.app, .apk, .ipa etc...) to TestObject server
+ */
+TestObject.uploadTestObjectApp = async function (app) {
+  if (!process.env.TESTOBJECT_API_KEY) {
+    throw new Error('Must provide TESTOBJECT_API_KEY in environment variables');
+  }
+  if (!process.env.TESTOBJECT_USERNAME) {
+    throw new Error('Must define TESTOBJECT_USERNAME in environment variables');
+  }
+
+  // If this app was already uploaded and hasn't changed since last upload, re-use the app id
+  const cache = TestObject._appIdCache[app];
+  if (cache) {
+    const {mtime} = await fs.stat(app);
+    const modifiedTime = +(new Date(mtime));
+    if (modifiedTime < cache.uploaded) {
+      return cache.id;
+    }
+  }
+
+  // Upload the app
+  // (see https://app.staging.testobject.org/#/danielgraham/contact-manager/appium/basic/instructions)
+  // TODO: Blocked on TestObject fixing this
+  /*const appId = await exec('curl', [
+    '-u', `"${process.env.TESTOBJECT_USERNAME}:${process.env.TESTOBJECT_API_KEY}"`,
+    '-X', 'POST', 'https://app.staging.testobject.org:443/api/storage/upload',
+    '-H', '"Content-Type: application/octet-stream"',
+    '---data-binary', `@${app}`,
+  ]);*/
+  const appId = null;
+
+  TestObject._appIdCache[app] = {
+    id: appId,
+    uploaded: +(new Date()),
+  };
+
+  return appId;
+};
+
+TestObject.MOCHA_TESTOBJECT_TIMEOUT = 10 * 60 * 1000; // Ten minutes
+
+TestObject.HOST = 'app.staging.testobject.org';
+TestObject.PORT = 443;
+
+/**
+ * Overrides the WD (admc/wd) prototype to force it to use TestObject
+ */
+TestObject.overrideWD = function (wd, appiumZipUrl) {
+  const originalPromiseChainRemote = wd.promiseChainRemote;
+
+  // Override the 'promiseChainRemote' method to use TestObject HOST and special caps
+  wd.promiseChainRemote = async function (...args) {
+    let testObjectParams = {
+      host: TestObject.HOST,
+      port: TestObject.PORT,
+      https: true,
+    };
+    let driver;
+    if (_.isObject(args[0])) {
+      testObjectParams = Object.assign(args[0], testObjectParams);
+    }
+    driver = await originalPromiseChainRemote(testObjectParams);
```

Comment:
I don't think so, but I might as well change it. It can't hurt to set the context to `wd`.